fix product dropdown closing when user clicks in search box

haszari opened the pull request:
## Summary
The product search dropdown closes when the user clicks in the search box. This is unusual behaviour for a search box, and was [reported as an issue against WooCommerce.com](https://github.com/Automattic/woocommerce.com/issues/4597). For example, it is frustrating if you click within your search query to fix a typo!

This PR adds a check to the click handler that closes (toggles) the dropdown. If the event was triggered on the search box, it returns early, so the dropdown stays open.

## Testing
- Get HappyChat up and running (I'm new to this repo .. `npm install` `npm start` 
http://localhost:9000 !?)
- Click "What product do you need help with?" dropdown.
- Test the dropdown to make sure it's functioning correctly: 
- Click in search field => dropdown should remain open
- Edit search query, click within query text => dropdown should remain open
- Clicking an item in the dropdown should still select it and close the dropdown

haszari commented:
I'm not particularly happy with how I've updated the tests for this change. I'll happily address any review feedback about the tests or anything else!

nosolosw commented on src/ui/components/select-dropdown/index.jsx:
```diff
@@ -289,7 +294,14 @@ class SelectDropdown extends Component {
 	}
 
-	toggleDropdown() {
+	onContainerClick( event ) {
```
Should we rename this to just `onClick`? The click's target can be the own SelectDropdown component, its parent, etc.

nosolosw commented on src/ui/components/select-dropdown/test/index.js:
```diff
@@ -39,15 +39,16 @@ describe( 'index', () => {
+			const originalOnContainerClick = SelectDropdown.prototype.onContainerClick;
+			SelectDropdown.prototype.onContainerClick = jest.fn();
 
+			const fakeEvent = prepareFakeClickEvent( { } );
```
Can we inline the `prepareFakeClickEvent` function? I don't see it used anywhere else.

nosolosw commented on src/ui/components/select-dropdown/test/index.js:
```diff
@@ -39,15 +39,16 @@ describe( 'index', () => {
 
-			expect( SelectDropdown.prototype.toggleDropdown.mock.calls.length ).toBe( 1 );
-			SelectDropdown.prototype.toggleDropdown = originalToggleDropdown;
+			expect( SelectDropdown.prototype.onContainerClick.mock.calls.length ).toBe( 1 );
```
We should check that the `toggleDropdown` method wasn't called to let tests catch any future regression. Also: we can add a new test to make sure the `toggelDropdown` method is called at some point. I guess we can do this by simulating a click on one of the parent divs.

haszari replied:
I hit a snag here, I'm not sure how to set this up correctly in jest – do you have experience with this, can you point me to an example? The problem is that the behaviour we want to test depends on the DOM state (to check where the click occurred).

It looks like there is [some support for mocking refs](https://reactjs.org/blog/2016/11/16/react-v15.4.0.html#mocking-refs-for-snapshot-testing), but it looks like it's focused more on snapshot testing. Maybe snapshot testing is a better fit here, since we're testing DOM changes – what do you think? 

Another option would be to factor out the DOM ref check, and unit test a method that takes a boolean `clickTargetIsSearchBox` parameter. I'm happy to do this if you think that's a good approach. 

I'm pretty new to testing in React and Jest (and snapshot testing!), so happy to take advice here :)

nosolosw replied:
As per enzyme's docs:  `ShallowWrapper.simulate` [won't simulate](http://airbnb.io/enzyme/docs/api/ShallowWrapper/simulate.html#common-gotchas) an actual DOM event, so we need to use their `Full DOM rendering.simulate` [API](http://airbnb.io/enzyme/docs/api/ReactWrapper/simulate.html) to test what we want.

nosolosw commented on src/ui/components/select-dropdown/index.jsx:
```diff
@@ -289,7 +294,14 @@ class SelectDropdown extends Component {
+	onContainerClick( event ) {
+		// don't toggle the dropdown if the user clicks in the search box
+		if ( event.target === ReactDom.findDOMNode( this.refs.dropdownSearchBox ) )
+			return;
```
Would it be a good idea to `preventDefault` the event so it isn't captured by any other target?

haszari replied:
I don't think so – this broke the little `x` button to clear the search. The main thing I'm focused on here is preventing `toggleDropdown()` from being called – it's not really "handling" the click.

nosolosw commented:
@haszari This is very exciting! Thanks for investing time to improve this client for everyone. That's the main reason it exists: to benefit from the network effects of having a shared codebase that can be deployed to several clients.

The PR works. Code-wise I left some comments that I'd like you to address/comment before merging this.

haszari commented:
@nosolosw Thanks for the review! I've addressed most of your feedback.

We still need to figure out the best way to test this behaviour – if you have suggestions here, I'll follow up, if not, when I get a chance I'll research more (Enzyme?) to find a good solution.

nosolosw commented:
@haszari I pushed some tests in 719e475. We seem to live in antipodean time zones, so I'll go ahead and  merge this so it's no longer a blocker. Thanks for your work!
